blockstore/badger/versions/badger_v4.go

```diff
@@ -0,0 +1,304 @@
+)
+
+// BadgerV4 wraps the Badger v4 database to implement the BadgerDB interface.
+type BadgerV4 struct {
+	*badger.DB
+}
+
+func (b *BadgerV4) Close() error {
+	return b.DB.Close()
+}
+
+func (b *BadgerV4) IsClosed() bool {
+	return b.DB.IsClosed()
+}
+
+func (b *BadgerV4) NewStream() BadgerStream {
+	return &BadgerV4Stream{b.DB.NewStream()}
+}
+
+func (b *BadgerV4) Update(fn func(txn Txn) error) error {
+	return b.DB.Update(func(txn *badger.Txn) error {
+		return fn(&BadgerV4Txn{txn})
+	})
+}
+
+func (b *BadgerV4) View(fn func(txn Txn) error) error {
+	return b.DB.View(func(txn *badger.Txn) error {
+		return fn(&BadgerV4Txn{txn})
+	})
+}
+
+func (b *BadgerV4) NewTransaction(update bool) Txn {
+	return &BadgerV4Txn{b.DB.NewTransaction(update)}
+}
+
+func (b *BadgerV4) RunValueLogGC(discardRatio float64) error {
+	return b.DB.RunValueLogGC(discardRatio)
+}
+
+func (b *BadgerV4) Sync() error {
+	return b.DB.Sync()
+}
+
+func (b *BadgerV4) MaxBatchCount() int64 {
+	return b.DB.MaxBatchCount()
+}
+
+func (b *BadgerV4) MaxBatchSize() int64 {
+	return b.DB.MaxBatchSize()
+}
+
+func (b *BadgerV4) IndexCacheMetrics() *ristretto.Metrics {
+	return b.DB.IndexCacheMetrics()
+}
+
+func (b *BadgerV4) GetErrKeyNotFound() error {
+	return badger.ErrKeyNotFound
+}
+
+func (b *BadgerV4) GetErrNoRewrite() error {
+	return badger.ErrNoRewrite
+}
+
+func (b *BadgerV4) NewWriteBatch() WriteBatch {
+	return &BadgerV4WriteBatch{b.DB.NewWriteBatch()}
+}
+
+func (b *BadgerV4) Flatten(workers int) error {
+	return b.DB.Flatten(workers)
+}
+
+func (b *BadgerV4) Size() (lsm int64, vlog int64) {
+	return b.DB.Size()
+}
+
+func (b *BadgerV4) Copy(ctx context.Context, to BadgerDB) (defErr error) {
+
+	batch := to.NewWriteBatch()
+	defer func() {
+		if defErr == nil {
+			defErr = batch.Flush()
+		}
+		if defErr != nil {
+			batch.Cancel()
+		}
+	}()
+
+	stream := b.DB.NewStream()
+
+	return iterateBadgerV4(ctx, stream, func(kvs []*pb.KV) error {
+		// check whether context is closed on every kv group
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		for _, kv := range kvs {
+			if err := batch.Set(kv.Key, kv.Value); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func iterateBadgerV4(ctx context.Context, stream *badger.Stream, iter func([]*pb.KV) error) error {
+	workers := IterateLSMWorkers
+	if workers == 0 {
+		workers = between(2, 8, runtime.NumCPU()/2)
+	}
+
+	stream.NumGo = workers
+	stream.LogPrefix = "iterateBadgerKVs"
+	stream.Send = func(buf *z.Buffer) error {
+		kvl, err := badger.BufferToKVList(buf)
+		if err != nil {
+			return fmt.Errorf("buffer to KV list conversion: %w", err)
+		}
+
+		kvs := make([]*pb.KV, 0, len(kvl.Kv))
+		for _, kv := range kvl.Kv {
+			if kv.Key != nil && kv.Value != nil {
+				kvs = append(kvs, kv)
+			}
+		}
+		if len(kvs) == 0 {
+			return nil
+		}
+		return iter(kvs)
+	}
+	return stream.Orchestrate(ctx)
+}
+
+func (b *BadgerV4) DefaultOptions(path string, readonly bool) Options {
+	var opts Options
```

Comment:
I'm a bit concerned that users will try to run badger v4 but then the difference in default options will make them think something is wrong.  I guess it depends on whether we think people should upgrade to latest version without being badger power users.  What was your thinking behind dropping all of the default settings for badger V2 for V4?

Reply:
this is an artifact of some confusion I had implementing the PR. I've cleaned it up and removed this code c4611608f. thanks for the call out 👍